File: collect_app/src/test/java/org/odk/collect/android/widgets/GeoTraceWidgetTest.java
package org.odk.collect.android.widgets;

import android.view.View;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.javarosa.core.model.data.StringData;
import org.javarosa.form.api.FormEntryPrompt;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.odk.collect.android.R;
import org.odk.collect.android.formentry.questions.QuestionDetails;
import org.odk.collect.android.geo.MapConfigurator;
import org.odk.collect.android.listeners.WidgetValueChangedListener;
import org.odk.collect.android.widgets.interfaces.GeoDataRequester;
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.odk.collect.android.widgets.support.GeoWidgetHelpers.stringFromDoubleList;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.mockValueChangedListener;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithAnswer;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnlyAndAnswer;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;

@RunWith(AndroidJUnit4.class)
public class GeoTraceWidgetTest {
    private final String answer = stringFromDoubleList();

    private WaitingForDataRegistry waitingForDataRegistry;
    private GeoDataRequester geoDataRequester;
    private MapConfigurator mapConfigurator;

    @Before
    public void setup() {
        waitingForDataRegistry = mock(WaitingForDataRegistry.class);
        geoDataRequester = mock(GeoDataRequester.class);
        mapConfigurator = mock(MapConfigurator.class);
        when(mapConfigurator.isAvailable(any())).thenReturn(true);
    }

    @Test
    public void getAnswer_whenPromptDoesNotHaveAnswer_returnsNull() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        assertNull(widget.getAnswer());
    }

    @Test
    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
        assertEquals(widget.getAnswer().getDisplayText(), answer);
    }

    @Test
    public void whenPromptDoesNotHaveAnswer_textViewDisplaysEmptyString() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        assertEquals(widget.binding.geoAnswerText.getText().toString(), "");
    }

    @Test
    public void whenPromptHasAnswer_textViewDisplaysAnswer() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
        assertEquals(widget.binding.geoAnswerText.getText().toString(), answer);
    }

    @Test
    public void whenPromptIsReadOnlyAndDoesNotHaveAnswer_geoButtonIsNotDisplayed() {
        GeoTraceWidget widget = createWidget(promptWithReadOnly());
        assertEquals(widget.binding.simpleButton.getVisibility(), View.GONE);
    }

    @Test
    public void whenPromptIsReadOnlyAndHasAnswer_viewGeoShapeButtonIsShown() {
        GeoTraceWidget widget = createWidget(promptWithReadOnlyAndAnswer(new StringData(answer)));
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.geotrace_view_read_only));
    }

    @Test
    public void whenPromptIsNotReadOnlyAndDoesNotHaveAnswer_startGeoShapeButtonIsShown() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.get_trace));
    }

    @Test
    public void whenPromptIsNotReadOnlyAndHasAnswer_viewOrChangeGeoShapeButtonIsShown() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.geotrace_view_change_location));
    }

    @Test
    public void clearAnswer_clearsWidgetAnswer() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
        widget.clearAnswer();

        assertEquals(widget.binding.geoAnswerText.getText(), "");
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.get_trace));
    }

    @Test
    public void clearAnswer_callsValueChangeListeners() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
        widget.clearAnswer();

        verify(valueChangedListener).widgetValueChanged(widget);
    }

    @Test
    public void clickingButtonAndAnswerTextViewForLong_callsLongClickListener() {
        View.OnLongClickListener listener = mock(View.OnLongClickListener.class);
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));

        widget.setOnLongClickListener(listener);
        widget.binding.simpleButton.performLongClick();
        widget.binding.geoAnswerText.performLongClick();

        verify(listener).onLongClick(widget.binding.simpleButton);
        verify(listener).onLongClick(widget.binding.geoAnswerText);
    }

    @Test
    public void setData_updatesWidgetAnswer() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        widget.setData(answer);
        assertEquals(widget.getAnswer().getDisplayText(), answer);
    }

    @Test
    public void setData_setsCorrectAnswerInAnswerTextView() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        widget.setData(answer);
        assertEquals(widget.binding.geoAnswerText.getText().toString(), answer);
    }

    @Test
    public void setData_updatesWidgetDisplayedAnswer() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        widget.setData(answer);
        assertEquals(widget.binding.geoAnswerText.getText().toString(), answer);
    }

    @Test
    public void setData_whenDataIsNull_updatesButtonLabel() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(new StringData(answer)));
        widget.setData("");
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.get_trace));
    }

    @Test
    public void setData_whenDataIsNotNull_updatesButtonLabel() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        widget.setData(answer);
        assertEquals(widget.binding.simpleButton.getText(), widget.getContext().getString(R.string.geotrace_view_change_location));
    }

    @Test
    public void setData_callsValueChangeListener() {
        GeoTraceWidget widget = createWidget(promptWithAnswer(null));
        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
        widget.setData(answer);

        verify(valueChangedListener).widgetValueChanged(widget);
    }

    @Test
    public void buttonClick_whenMapConfiguratorIsUnavailable_doesNotRequestGeoTrace() {
        FormEntryPrompt prompt = promptWithAnswer(null);
        GeoTraceWidget widget = createWidget(prompt);

        when(mapConfigurator.isAvailable(widget.getContext())).thenReturn(false);
        widget.binding.simpleButton.performClick();

        verify(geoDataRequester, never()).requestGeoTrace(widget.getContext(), prompt, "",  waitingForDataRegistry);
        verify(mapConfigurator).showUnavailableMessage(widget.getContext());
    }

    @Test
    public void buttonClick_whenMapConfiguratorIsAvailable_requestsGeoTrace() {
        FormEntryPrompt prompt = promptWithAnswer(null);
        GeoTraceWidget widget = createWidget(prompt);
        widget.binding.simpleButton.performClick();

        verify(geoDataRequester).requestGeoTrace(widget.getContext(), prompt, "", waitingForDataRegistry);
    }

    @Test
    public void buttonClick_requestsGeoTrace_whenAnswerIsCleared() {
        FormEntryPrompt prompt = promptWithAnswer(new StringData(answer));
        GeoTraceWidget widget = createWidget(prompt);
        widget.clearAnswer();
        widget.binding.simpleButton.performClick();

        verify(geoDataRequester).requestGeoTrace(widget.getContext(), prompt, "", waitingForDataRegistry);
    }

    @Test
    public void buttonClick_requestsGeoTrace_whenAnswerIsUpdated() {
        FormEntryPrompt prompt = promptWithAnswer(null);
        GeoTraceWidget widget = createWidget(prompt);
        widget.setData(answer);
        widget.binding.simpleButton.performClick();

        verify(geoDataRequester).requestGeoTrace(widget.getContext(), prompt, answer, waitingForDataRegistry);
    }

    private GeoTraceWidget createWidget(FormEntryPrompt prompt) {
        return new GeoTraceWidget(widgetTestActivity(), new QuestionDetails(prompt),
                waitingForDataRegistry, mapConfigurator, geoDataRequester);
    }
}

File: collect_app/src/test/java/org/odk/collect/android/widgets/GeoTraceWidgetTest.kt
package org.odk.collect.android.widgets

import android.app.Application
import android.content.ComponentName
import android.graphics.Bitmap
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.javarosa.core.model.Constants
import org.javarosa.core.model.data.StringData
import org.javarosa.form.api.FormEntryPrompt
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.support.MockFormEntryPromptBuilder
import org.odk.collect.android.support.WidgetTestActivity
import org.odk.collect.android.widgets.interfaces.GeoDataRequester
import org.odk.collect.android.widgets.support.GeoWidgetHelpers.stringFromDoubleList
import org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithAnswer
import org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly
import org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnlyAndAnswer
import org.odk.collect.androidtest.onNodeWithClickLabel
import org.odk.collect.maps.MapPreviewRenderer
import org.odk.collect.maps.traces.TraceDescription
import org.odk.collect.strings.R.string
import org.robolectric.Shadows.shadowOf

@RunWith(AndroidJUnit4::class)
class GeoTraceWidgetTest {
    init {
        shadowOf(getApplicationContext<Application>().packageManager)
            .addActivityIfNotPresent(ComponentName(getApplicationContext(), WidgetTestActivity::class.java))
    }

    @get:Rule
    val composeRule = createAndroidComposeRule<WidgetTestActivity>()

    private val answer = stringFromDoubleList()

    private val geoDataRequester = mock<GeoDataRequester>()

    private val mapPreviewRenderer = FakeMapPreviewRenderer()

    @Test
    fun `#getAnswer returns null when there is no answer`() {
        val widget = createWidget(promptWithAnswer(null))
        assertThat(widget.answer, equalTo(null))
    }

    @Test
    fun `#getAnswer returns the answer when there is one`() {
        val widget = createWidget(promptWithAnswer(StringData(answer)))
        assertThat(widget.answer!!.displayText, equalTo(answer))
    }

    @Test
    fun `the answer is not displayed when there is no answer`() {
        createWidget(promptWithAnswer(null))
        composeRule.onNodeWithText(answer).assertDoesNotExist()
    }

    @Test
    fun `the answer is displayed when there is one`() {
        createWidget(promptWithAnswer(StringData(answer)))
        composeRule.onNodeWithText(answer).assertIsDisplayed()
    }

    @Test
    fun `the map preview is displayed instead of the answer when it can be rendered`() {
        mapPreviewRenderer.preview = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)
        createWidget(promptWithAnswer(StringData(answer)))
        composeRule.onNodeWithTag(MAP_PREVIEW_TAG).assertIsDisplayed()
        composeRule.onNodeWithText(answer).assertDoesNotExist()
    }

    @Test
    fun `the answer is displayed instead of the map preview when it can't be rendered`() {
        mapPreviewRenderer.preview = null
        createWidget(promptWithAnswer(StringData(answer)))
        composeRule.onNodeWithTag(MAP_PREVIEW_TAG).assertDoesNotExist()
        composeRule.onNodeWithText(answer).assertIsDisplayed()
    }

    @Test
    fun `the button is hidden when the question is read-only and there is no answer`() {
        createWidget(promptWithReadOnly())
        composeRule.onNodeWithClickLabel(string.view_line).assertDoesNotExist()
    }

    @Test
    fun `the button has correct text when the question is read-only and there is an answer`() {
        createWidget(promptWithReadOnlyAndAnswer(StringData(answer)))
        composeRule.onNodeWithClickLabel(string.view_line).assertIsDisplayed()
    }

    @Test
    fun `the button has correct text when the question is not read-only and there is no answer`() {
        createWidget(promptWithAnswer(null))
        composeRule.onNodeWithClickLabel(string.get_line).assertIsDisplayed()
    }

    @Test
    fun `the button has correct text when the question is not read-only and there is an answer`() {
        createWidget(promptWithAnswer(StringData(answer)))
        composeRule.onNodeWithClickLabel(string.view_or_change_line).assertIsDisplayed()
    }

    @Test
    fun `clicking the button requests a geotrace`() {
        val prompt = promptWithAnswer(null)
        createWidget(prompt)

        composeRule.onNodeWithClickLabel(string.get_line).performClick()

        verify(geoDataRequester).requestGeoPoly(prompt)
    }

    @Test
    fun `clicking the button requests a geotrace when there is an answer`() {
        val prompt = promptWithAnswer(StringData(answer))
        createWidget(prompt)

        composeRule.onNodeWithClickLabel(string.view_or_change_line).performClick()

        verify(geoDataRequester).requestGeoPoly(prompt)
    }

    private fun createWidget(prompt: FormEntryPrompt) = GeoTraceWidget(
        composeRule.activity,
        QuestionDetails(asGeoTraceQuestion(prompt)),
        geoDataRequester,
        QuestionWidget.Dependencies(mock(), MediaWidgetAnswerViewModel(mock(), mock(), mock(), mapPreviewRenderer))
    ).also {
        composeRule.activity.setContentView(it)
    }

    private fun asGeoTraceQuestion(prompt: FormEntryPrompt) = MockFormEntryPromptBuilder(prompt)
        .withControlType(Constants.CONTROL_INPUT)
        .withDataType(Constants.DATATYPE_GEOTRACE)
        .build()

    private class FakeMapPreviewRenderer : MapPreviewRenderer {
        var preview: Bitmap? = null

        override fun render(
            trace: TraceDescription,
            width: Int,
            height: Int,
            callback: (Bitmap?) -> Unit
        ): () -> Unit {
            callback(preview)
            return {}
        }
    }
}
